src/DotVVM.Framework.Tests.Common/ResourceManagement/ResourceRepositoryTests.cs

```diff
@@ -25,5 +25,199 @@ public void ResourceRepository_CyclicDependency_Throws()
                 manager.AddRequiredResource("one");
             });
         }
+
+        [TestMethod]
+        public void ResourceRepository_CyclicDependency_Structure_1()
+        {
+            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "6", "5" } });
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "4", "3" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "4", "3" } });
+
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "2", "1" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "1" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "dotvvm" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("7");
+            });
+        }
+
+        [TestMethod]
+        public void ResourceRepository_CyclicDependency_Structure_2_Throws()
+        {
+            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "6", "5" } });
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "4", "3" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "4", "3" } });
+
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "2", "1" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "1" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "DotVVM" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "DotVVM" } });
+
+                configuration.Resources.Register("DotVVM", new NullResource { Dependencies = new[] { "7" } }); ;//cyclic
```

Comment:
I don't think it's a good idea to name test resources `DotVVM`.

Reply:
@OndrejSulc:
> 2) renamed to "dotvvm"

come on... What I meant it that the resource name should not be the same as another resource that already exists. If you mean to test something specific about the case when the resource already exists, please create another specific test for that, or at least add a short explanation what should happen.